File: src/elixir/svd/lib/svd/validators/standardised_vessel_dataset_validator.ex
defmodule SVD.Validators.StandardisedVesselDatasetValidator do
  @moduledoc false

  alias SVD.Models
  alias SVD.Validators.{GeneralInformationValidator, ValidationError}

  @spec validate(term()) :: [ValidationError.t()]
  def validate(nil) do
    [
      %ValidationError{field: "SVD", message: "SVD cannot be nil"}
    ]
  end

  def validate(dataset) do
    case Models.normalize_dataset(dataset) do
      nil ->
        validate(nil)

      normalized ->
        errors = []

        if is_nil(normalized.general) do
          errors ++
            [
              %ValidationError{
                field: "General",
                message: "General information is required"
              }
            ]
        else
          errors ++ GeneralInformationValidator.validate(normalized.general)
        end
    end
  end
end

File: src/typescript/src/validators/StandardisedVesselDatasetValidator.ts
import { asObject, readProperty } from "../runtime/types";
import { GeneralValidator } from "./GeneralValidator";
import { ValidationError } from "./ValidationError";

export class StandardisedVesselDatasetValidator {
  private readonly generalValidator: GeneralValidator;

  public constructor(generalValidator: GeneralValidator = new GeneralValidator()) {
    this.generalValidator = generalValidator;
  }

  public validate(dataset: unknown): ValidationError[] {
    const datasetObject = asObject(dataset);
    if (datasetObject === null) {
      return [{ field: "SVD", message: "SVD cannot be nil", value: dataset }];
    }

    const general = readProperty(datasetObject, "general");
    const generalObject = asObject(general);
    if (generalObject === null) {
      return [{ field: "General", message: "General is required", value: general }];
    }

    return this.generalValidator.validate(generalObject);
  }
}
